rust-lang/rust-analyzer
# Add generic parameter assist

Closes #12531

# Supported features
Add a generic type parameter to anything that can take one. This includes struct, enum, union, fn, trait, type alias, trait alias, use alias.

Note that this produces code that does not compile, as the generic parameter must be used, and the assist does not add a member to the struct/enum/etc.

All of these examples are listed in the tests section at the bottom of the assist file (crates/ide-assists/src/handlers/add_generic_parameter.rs), but are some duplicated here for exposition:

Basic nesting (works same for enums, unions, traits)
```rust
struct Channel$0(u8);
struct Color([Channel; 3]);
struct Image(Vec<Color>);
```
```rust
struct Channel<T1>(u8);
struct Color<T1>([Channel<T1>; 3]);
struct Image<T1>(Vec<Color<T1>>);
```
It works for cyclic structs/etc, and adds the parameter on impls.
```rust
struct Cons$0(i32, List);
impl Cons{ fn len(&self){42} }
struct List(Option<Box<Cons>>);
impl List{ fn len(&self){42} }
```
```rust
struct Cons<T1>(i32, List<T1>);
impl<T1> Cons<T1>{ fn len(&self){42} }
struct List<T1>(Option<Box<Cons<T1>>>);
impl<T1> List<T1>{ fn len(&self){42} }
```
For traits, it adds the generic to the trait, rather than the function. This seems to generally be the desired outcome
```rust
trait HasFood$0 {}
trait CanMakeFood{ fn make() -> impl HasFood; }
```
```rust
trait HasFood<T1> {}
trait CanMakeFood<T1>{ fn make() -> impl HasFood<T1>; }
```
For impls, it uses the impl's generic, but if there is none, it adds to the innermost (likely a function)
```rust
struct Color$0(u8);
struct MakeColor;
impl MakeColor { fn make() -> Color { Color(0) }}
struct MakeColor2(Color);
impl MakeColor2 { fn make() -> Color { Color(0) }}
```
```rust
struct Color<T1>(u8);
struct MakeColor;
impl MakeColor { fn make<T1>() -> Color<T1> { Color(0) }}
struct MakeColor2<T1>(Color<T1>);
impl<T1> MakeColor2<T1> { fn make() -> Color<T1> { Color(0) }}
```
Use aliases:
```rust
struct ColorImpl$0(u8);
use ColorImpl as Color;
struct Image(Vec<Color>);
```
```rust
struct ColorImpl<T1>(u8);
use ColorImpl as Color;
struct Image<T1>(Vec<Color<T1>>);
```
Type/trait aliases:
```rust
trait Debug2$0 = Debug;
trait SuperDebug: Debug2 {}
```
```rust
trait Debug2<T1> = Debug;
trait SuperDebug<T1>: Debug2<T1> {}
```
More complicated interaction through a trait:
```rust
struct Food$0;
trait MakeFood{ fn make() -> Food; }
impl MakeFood for Farm{ fn make() -> Food { Food }}
```
```rust
struct Food<T1>;
trait MakeFood<T1>{ fn make() -> Food<T1>; }
impl<T1> MakeFood<T1> for Farm{ fn make() -> Food<T1> { Food }}
```


I've tested this on toy AST's and it works fine, but I haven't tried larger scenarios. Larger things might try to modify things (e.g. traits) that aren't owned by the user, which could be problematic.

# Showcase
- none, look at examples above, not much interactivity (yet).

# Future
- I'd like to have the user pick the generic parameter name. I just use `T1` right now.
- I haven't implemented solving name collisions, like the following, which would add a duplicate the `T1` parameter:
	- 
        ```rust
		struct Thing;
        trait DoStuff{ fn stuff<T1>(hello: Thing$0); }
		```
- I'd like to add support for bounds and more kinds of generics (const, lifetime), but it is unclear how to add it to the editor/lsp, read below.
	- [edit 2023/08/25] Possible implementation: 
        ```rust
		// show assist for any part of this attribute
		// note: this attribute isn't defined, its just for the assist to look at
		#[assist::add_generic_parameters$0]
		struct T<'a, T, const N: usize>;
		// will add T's generics to Thing (possibly deleting struct T)
		struct Thing;
		```
		- a similar strategy could be used in other assists for more advanced things:
			- apply assist for all descendant nodes applicable
			- change fn signature
			- move definition to some module
		
- test for edge cases since many different kinds of things can have generics and interact with each other.
- Is there anything I'm not seeing that could be improved, changed?
- I'd like some external direction in how to workflow, contribute better/easier, optimize, clean up, etc.

# About the coding

I have a few difficulty/uncertainty points when making this:

- rust-analyzer questions
	- syntax/ast api:
		- confusing when they get invalidated: if you modify a previous ast node, then it makes a later ast node have invalid offsets, but only if you don't make_mut on the later node
		- when to call clone_for_update/clone_subtree/make_mut
		- My solution was to buffer up everything I wanted to make_mut, then make_mut and modify.
	- There were a few missing pieces in the library code, do I add it? e.g. a trait impl that seemed obvious (I assumed so, so i added them)
	- Should I pull request a mostly done thing? are fixmes fine? if they're fine, how many are allowed?

- general assist questions:
	- input fuzzer? check for panic'ing inputs
	- how to do multi-file input tests
	- error reporting
		- silently fail?
		- print an error somewhere?
	- Is there a way to add more powerful assists (require more user input)?
		- make a command?
		- powerful assist examples: 
			- "extract (possibly multiple) selected expressions as parameters"
			- "add multiple generic parameters, with bounds: <'a, T: Clone, const N: usize>"
			- "reorder function parameters (x,y,z) -> (z,x,y)"

- how are small details discussed? 
	- when assist should be applicable
	- assist specific details
		- placeholder names (e.g. have `T1` be a placeholder selected by the user after assist execution)
		- I have a bunch of these in FIXME comments

- specific to add_generic_parameter:
	- customize the parameter name? snippet?
	- many options: lifetime, const
		- can these be in a menu?
	- remove generic parameter?
	- what should Assist.target be? (Right now it is the entire file, since this is as far reaching as a rename, which is pretty far)
